### PR TITLE
Add support for pre aggregated metrics

### PR DESCRIPTION
**Pull request by hectorhdzg**
Including following metrics:
Dependency duration
Server response time
Exceptions count
Traces count

### CODE REVIEW

**Comment by hectorhdzg** [comment]
Still working on tests

File: AutoCollection/PreAggregatedMetrics.ts [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+
+    public static countRequest(duration: number | string, dimensions: MetricRequestDimensions) {
+        if (!AutoCollectPreAggregatedMetrics.isEnabled()) {
+            return;
+        }
+        let durationMs: number;
+        let counter: AggregatedMetricCounter = AutoCollectPreAggregatedMetrics._getAggregatedCounter(dimensions, this._requestCountersCollection);
+        if (typeof duration === 'string') {
+            // dependency duration is passed in as "00:00:00.123" by autocollectors
+            durationMs = +new Date('1970-01-01T' + duration + 'Z'); // convert to num ms, returns NaN if wrong
```
Comment (MSNev) [inline AutoCollection/PreAggregatedMetrics.ts:106]: Do (because caller is controlled by this code) or should we handle invalid values for strings?

Comment (hectorhdzg) [reply AutoCollection/PreAggregatedMetrics.ts:106]: The string case is only for internal code, basically is receiving the duration string included in the generated envelope

**Review by MSNev — APPROVED** [review]

File: AutoCollection/PreAggregatedMetrics.ts [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+    private static _client: TelemetryClient;
+    private _handle: NodeJS.Timer;
+    private _isEnabled: boolean;
+    private _isInitialized: boolean;
+    private static _dependencyCountersCollection: Array<AggregatedMetricCounter>;
+    private static _requestCountersCollection: Array<AggregatedMetricCounter>;
+    private static _exceptionCountersCollection: Array<AggregatedMetricCounter>;
+    private static _traceCountersCollection: Array<AggregatedMetricCounter>;
+
+    constructor(client: TelemetryClient, collectionInterval = 60000) {
```
Comment (lzchen) [inline AutoCollection/PreAggregatedMetrics.ts:31]: Is this interval in ms?

File: TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+import Context = require("../Library/Context");
+
+export function preAggregatedMetricsTelemetryProcessor(envelope: Contracts.EnvelopeTelemetry, context: Context): boolean {
+    // Increment rate counters
+    switch (envelope.data.baseType) {
+        case TelemetryType.TelemetryTypeString.Exception:
+            let exceptionDimensions: MetricExceptionDimensions = {
+                cloudRoleInstance: envelope.tags[context.keys.cloudRoleInstance],
+                cloudRoleName: envelope.tags[context.keys.cloudRole],
+            };
```
Comment (lzchen) [inline TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts:19]: `exceptions/count` also specifies a `client/type` field in the required dimensions. Is that needed here? What is dot net sdk doing?

File: TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+            AutoCollecPreAggregatedMetrics.countException(exceptionDimensions);
+            break;
+        case TelemetryType.TelemetryTypeString.Trace:
+            let traceDimensions: MetricTraceDimensions = {
+                cloudRoleInstance: envelope.tags[context.keys.cloudRoleInstance],
+                cloudRoleName: envelope.tags[context.keys.cloudRole],
```
Comment (lzchen) [inline TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts:25]: `trace/count` requires dimensions `trace/severityLevel` and  `operation/synthetic` as well.

File: applicationinsights.ts [hunk 5]
```diff
@@ -242,6 +244,20 @@ export class Configuration {
         return Configuration;
     }
 
+    /**
+     * Sets the state of pre aggregated metrics tracking (enabled by default)
+     * @param value if true pre aggregated metrics will be collected every minute and sent to Application Insights
+     * @returns {Configuration} this class
+     */
+    public static setAutoCollectPreAggregatedMetrics(value: boolean) {
```
Comment (lzchen) [inline applicationinsights.ts:252]: Curious, this api is exposed to users to enable/disable the feature?

File: AutoCollection/PreAggregatedMetrics.ts [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+    public enable(isEnabled: boolean, collectionInterval?: number) {
+        this._isEnabled = isEnabled;
+        if (this._isEnabled && !this._isInitialized) {
+            this._isInitialized = true;
+        }
+
+        if (isEnabled) {
+            if (!this._handle) {
+                for (let i = 0; i < AutoCollectPreAggregatedMetrics._dependencyCountersCollection.length; i++) {
+                    AutoCollectPreAggregatedMetrics._dependencyCountersCollection[i].lastTotalCount = AutoCollectPreAggregatedMetrics._dependencyCountersCollection[i].totalCount;
```
Comment (lzchen) [inline AutoCollection/PreAggregatedMetrics.ts:54]: What does this line do? Set everything to 0?

File: AutoCollection/PreAggregatedMetrics.ts [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+
+class AutoCollectPreAggregatedMetrics {
+
+    public static INSTANCE: AutoCollectPreAggregatedMetrics;
+    private _collectionInterval: number;
+    private static _client: TelemetryClient;
+    private _handle: NodeJS.Timer;
+    private _isEnabled: boolean;
+    private _isInitialized: boolean;
+    private static _dependencyCountersCollection: Array<AggregatedMetricCounter>;
```
Comment (lzchen) [inline AutoCollection/PreAggregatedMetrics.ts:26]: I'm wondering why you need to a keep a collection of counters? We only need one counter per request type.

File: AutoCollection/PreAggregatedMetrics.ts [hunk 8]
```diff
@@ -0,0 +1,301 @@
...
+        // Build metric properties
+        let metricProperties: any = {};
+        for (let dim in metric.dimensions) {
+            metricProperties[PreaggregatedMetricPropertyNames[dim as MetricDimensionTypeKeys]] = metric.dimensions[dim];
+        }
+        metricProperties = {
+            ...metricProperties,
+            "_MS.MetricId": metric.metricType,
+            "_MS.AggregationIntervalMs": String(metric.aggregationInterval),
+            "_MS.IsAutocollected": "True",
```
Comment (lzchen) [inline AutoCollection/PreAggregatedMetrics.ts:281]: You need to populate `Request.Success` = `True/False` depending on success or not for requests/dependencies failed. This is in ADDITION to `request/success` and `dependencies/success` dimensions.

File: TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts (hunk 3 above)
Comment (hectorhdzg) [reply TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts:19]: .Net doesn't send this dimension in pre aggregated metrics and we don't support that property in node.js SDK at all, is this something available in other SDKs?

File: AutoCollection/PreAggregatedMetrics.ts (hunk 2 above)
Comment (hectorhdzg) [reply AutoCollection/PreAggregatedMetrics.ts:31]: Yes, will add documentation for it

File: TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    MetricDependencyDimensions,
+    MetricExceptionDimensions,
+    MetricRequestDimensions,
+    MetricTraceDimensions
+} from "../Declarations/Metrics/AggregatedMetricDimensions";
+import Context = require("../Library/Context");
+
+export function preAggregatedMetricsTelemetryProcessor(envelope: Contracts.EnvelopeTelemetry, context: Context): boolean {
+    // Increment rate counters
+    switch (envelope.data.baseType) {
```
Comment (lzchen) [inline TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts:14]: You need to add the `_MS.ProcessedByMetricExtractors` custom dimension to these to indicate to breeze to not aggregate.

File: applicationinsights.ts (hunk 5 above)
Comment (hectorhdzg) [reply applicationinsights.ts:252]: Yes that is why this API is exposed, similar approach is used for other auto collection features

File: AutoCollection/PreAggregatedMetrics.ts (hunk 7 above)
Comment (hectorhdzg) [reply AutoCollection/PreAggregatedMetrics.ts:26]: AggregatedMetricCounter here is a class that have all necessary counters plus the dimensions, we need a collection to send all data by dimensions

**Review by hectorhdzg — COMMENTED** [review]
Tests are available now

File: TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts (hunk 3 above)
Comment (lzchen) [reply TelemetryProcessors/PreAggregatedMetricsTelemetryProcessor.ts:24]: If .Net doesn't send this it's fine to leave it out.

**Review by lzchen — APPROVED** [review]

**Pull request merged**